In [1]:
import csv, json, math, copy
import pandas as pd

load = pd.read_csv('data.csv')
print(load)

     planet         key                value  value~
0     Earth        core                 Iron     NaN
1     Earth        core               Nickel     NaN
2     Earth        core              Silicon     NaN
3     Earth        core               Sulfur     NaN
4     Earth  atmosphere             Nitrogen   78.00
5     Earth  atmosphere               Oxygen   21.00
6     Earth  atmosphere                Argon    1.00
7     Earth        mass                 1 Me     NaN
8     Earth    distance   8.32 light minutes     NaN
9      Mars        core                 Iron     NaN
10     Mars        core               Sulfur     NaN
11     Mars  atmosphere       Carbon Dioxide   95.32
12     Mars  atmosphere             Nitrogen    2.70
13     Mars  atmosphere                Argon    1.60
14     Mars        mass             0.107 Me     NaN
15     Mars    distance  12.67 light minutes     NaN
16  Mercury        core                 Iron     NaN
17  Mercury        core     Solid Inner Core  

In [2]:
columns = ['planet', 'key', 'value', 'value~']
fin_dict = dict()

for item in list(load[columns[0]]): # the planets
    fin_dict[item] = dict()
for index, row in load.iterrows():
    fin_dict[row[columns[0]]][row[columns[1]]]=[]
for index, row in load.iterrows():
    if row[columns[1]]=='atmosphere':
        if math.isnan(row[columns[3]])==False:
            val = {row[columns[2]]:row[columns[3]]}
        else:
            val = {row[columns[2]]:''}
    else:
        val = row[columns[2]]
    fin_dict[row[columns[0]]][row[columns[1]]].append(val)
    

In [3]:
print(fin_dict) #this is what should go to App to display in "answer search" mode

{'Earth': {'core': ['Iron', 'Nickel', 'Silicon', 'Sulfur'], 'atmosphere': [{'Nitrogen': 78.0}, {'Oxygen': 21.0}, {'Argon': 1.0}], 'mass': ['1 Me'], 'distance': ['8.32 light minutes']}, 'Mars': {'core': ['Iron', 'Sulfur'], 'atmosphere': [{'Carbon Dioxide': 95.32}, {'Nitrogen': 2.7}, {'Argon': 1.6}], 'mass': ['0.107 Me'], 'distance': ['12.67 light minutes']}, 'Mercury': {'core': ['Iron', 'Solid Inner Core'], 'atmosphere': [{'Hydrogen': ''}, {'Helium': ''}, {'Oxygen': ''}], 'distance': ['3.22 light minutes'], 'mass': ['0.055 Me']}, 'Venus': {'core': ['Iron', 'Liquid Core'], 'atmosphere': [{'Carbon Dioxide': 96.0}, {'Nitrogen': 3.0}], 'distance': ['6.02 light minutes'], 'mass': ['0.815 Em']}}


# Questions

## Core or Atmosphere
### Inclusion: "Which of these is in [planet]'s [core/atmosphere]?"
Generate two false from set of metals/gasses, include one from data

### Exclusion: "Which of these is NOT in [planet]'s [core/atmosphere]?"
Generate one false from set of metals/gasses, include two from data

## Distance
### Distance (planet unknown): "Which planet is [8.32 light minutes] from the sun?"
Generate two false planets from set of planets, include ans

### Distance (planet known): "How far away is [planet] from the sun?"
Generate two false distances, include ans

## Mass
### Mass (planet unknown): "Which planet is [0.815] times the mass of the earth?"
Generate two false planets, include ans

### Mass (planet known): "What is [planet]'s mass?"
Generate two false masses, include ans

In [4]:
false = { 'core': ['Gold', 'Silver', 'Aluminum'],
'atmosphere' : ['Xenon', 'Radon', 'Chlorine', 'Flourine'], 
'distance'  : ['7,047 light minutes', '2 million light minutes', '1.2 light minutes', '5.01527 light minutes'], 
'mass' : ['200 Me', '6.89 Me', '12.2 Me', '.0001 Me']}

dict_with_false = copy.deepcopy(fin_dict)
for i in list(dict_with_false.keys()): #planets
    for j in list(dict_with_false[i].keys()): #core, atmos, dist, mass
        new_key = j + '_false'
        dict_with_false[i][new_key] = false[j]
        
print(dict_with_false)

print(fin_dict)

{'Earth': {'core': ['Iron', 'Nickel', 'Silicon', 'Sulfur'], 'atmosphere': [{'Nitrogen': 78.0}, {'Oxygen': 21.0}, {'Argon': 1.0}], 'mass': ['1 Me'], 'distance': ['8.32 light minutes'], 'core_false': ['Gold', 'Silver', 'Aluminum'], 'atmosphere_false': ['Xenon', 'Radon', 'Chlorine', 'Flourine'], 'mass_false': ['200 Me', '6.89 Me', '12.2 Me', '.0001 Me'], 'distance_false': ['7,047 light minutes', '2 million light minutes', '1.2 light minutes', '5.01527 light minutes']}, 'Mars': {'core': ['Iron', 'Sulfur'], 'atmosphere': [{'Carbon Dioxide': 95.32}, {'Nitrogen': 2.7}, {'Argon': 1.6}], 'mass': ['0.107 Me'], 'distance': ['12.67 light minutes'], 'core_false': ['Gold', 'Silver', 'Aluminum'], 'atmosphere_false': ['Xenon', 'Radon', 'Chlorine', 'Flourine'], 'mass_false': ['200 Me', '6.89 Me', '12.2 Me', '.0001 Me'], 'distance_false': ['7,047 light minutes', '2 million light minutes', '1.2 light minutes', '5.01527 light minutes']}, 'Mercury': {'core': ['Iron', 'Solid Inner Core'], 'atmosphere': [{'H

In [87]:
question_types = ['core_in', 'core_ex', 'atmosphere_in', 'atmosphere_ex', 'distance_k','distance_uk', 'mass_u', 'mass_uk']
# question = base:'', inputs = []
# answers = display:[], answer:''

bases = {'core_in': 'Which of these is in {}\'s core?', \
         'core_ex': 'Which of these is NOT in {}\'s core?', \
         'atmosphere_in': 'Which of these is in {}\'s atmosphere?', \
         'atmosphere_ex': 'Which of these is NOT in {}\'s atmosphere?', \
#          'atmosphere_per': 'What percent of {}\'s atmophere is {}?',\
         'distance_k': 'How far away is {} from the sun?', \
         'distance_uk': 'Which planet is {} from the sun?', \
         'mass_k': 'What is {}\'s mass?',\
         'mass_uk': 'Which planet has mass {}?'
        }

planets = list(fin_dict.keys())
distances = []
masses = []
atmospheres = [] #includes percentages
gasses = [] #doesn't
cores = []

for planet in planets:
    distances.append(fin_dict[planet]['distance'][0])
    masses.append(fin_dict[planet]['mass'][0])
    atmospheres.append([planet, fin_dict[planet]['atmosphere']])
    cores.append(fin_dict[planet]['core'])

# for tup in atmospheres:
#     gas_list = tup[1]
#     for gas in gas_list:
#         gasses.append(list(gas.keys())[0])
# gasses = set(gasses)
    
# print(gasses)

inputs = {'core_in': planets, \
         'core_ex': planets, \
         'atmosphere_in': planets, \
         'atmosphere_ex': planets, \
#          'atmosphere_per': atmospheres, \
         'distance_k': planets, \
         'distance_uk': distances, \
         'mass_k': planets,\
         'mass_uk': masses
         }

outputs = {'core_in': 'core', \
         'core_ex': 'core', \
         'atmosphere_in': 'atmosphere', \
         'atmosphere_ex': 'atmosphere', \
#          'atmosphere_per': 'atmosphere', \
         'distance_k': 'distance', \
         'distance_uk': 'planet', \
         'mass_k': 'mass',\
         'mass_uk': 'planet'
         }

question_pieces = dict()
all_questions = dict()
for key in list(bases.keys()): #question types
    if key not in list(all_questions.keys()):
        all_questions[key] = []
    question_pieces[key] = {'base':bases[key]}
    question_pieces[key]['input'] = inputs[key]
    question_pieces[key]['output'] = outputs[key]
    question_pieces[key]['questions'] = []
    #for the different variations of what can be inputted, create a question for each:
    if key[-2:]=='uk':
        output_type = question_pieces[key]['output']
        for item in question_pieces[key]['input']:
            question = question_pieces[key]['base'].format(item)
            choices = []
            for k in list(dict_with_false.keys()):
                if dict_with_false[k][key[:-3]][0] == item:
                    ans = k
                    false_options = [i for i in planets if i!=ans]
                    for i in false_options:
                        for j in false_options:
                            if i!=j:
                                choice = [ans, i, j]
                                choices.append(choice)
                                question_dict = {'question':question, 'options': choice, 'answer': ans}
                                all_questions[key].append(question_dict)             
    else:
        for item in question_pieces[key]['input']:
#             if type(item) == list:
#                 pass
#                 gas_options = []
#                 for dic in item[1]:
#                     gas_options.append(list(dic.keys()))
#                 for gas in gas_options:
#                     question = question_pieces[key]['base'].format(item[0], gas)
            if type(item) == str:
                question = question_pieces[key]['base'].format(item)
                output_type = question_pieces[key]['output']
                ans_options = dict_with_false[item][output_type]
                false_options = dict_with_false[item][output_type+'_false']
                if key[-2:] == 'ex':
                    num_false = 1
                    one_set = false_options
                    two_set = ans_options
                else:
                    num_false = 2
                    one_set = ans_options
                    two_set = false_options
                choices = []
                for ans in one_set:
                    for i in two_set:
                        for j in two_set:
                            if i!=j:
                                choice = [ans, i, j]
                                if key == 'atmosphere_in' or key == 'atmosphere_ex':
                                    copy_choice = [ans,i,j]
                                    for c in choice:
                                        if type(c)==dict:
                                            copy_choice.remove(c)
                                            copy_choice.append(list(c.keys())[0])
                                    choice = copy_choice
                                question_dict = {'question':question, 'options': choice, 'answer': ans}
                                all_questions[key].append(question_dict) 
                            choices.append(choice)            

                question_pieces[key]['questions'].append(question)


print(all_questions)
with open('data.json', 'w') as outfile:
    json.dump(all_questions, outfile)

    

{'core_in': [{'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Gold', 'Silver'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Gold', 'Aluminum'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Silver', 'Gold'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Silver', 'Aluminum'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Aluminum', 'Gold'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Iron', 'Aluminum', 'Silver'], 'answer': 'Iron'}, {'question': "Which of these is in Earth's core?", 'options': ['Nickel', 'Gold', 'Silver'], 'answer': 'Nickel'}, {'question': "Which of these is in Earth's core?", 'options': ['Nickel', 'Gold', 'Aluminum'], 'answer': 'Nickel'}, {'question': "Which of these is in Earth's core?", 'options': ['Nickel', 'Silver', 'Gold'], '

In [44]:
# def get_choices(question_type, planet):
    
    
    
    
# def get_ans(question_type, planet):